src/models/tree.ts

```diff
@@ -155,11 +155,23 @@ export const Tree = types.model("Tree", {
                 return;
             }
             applySnapshot(model, snapshot);
+
+            // The contract is that the promise we return should not resolve
+            // until all of the changes have been applied to shared model.
+            // We should not wait until the tree has run 
+            // updateTreeAfterSharedModelChanges
+            // So we can just resolve immediately
+            return Promise.resolve();
```

Comment:
With all of these functions that are returning `Promise.resolve()`, I'd be inclined to mark them `async` and then put the comment about the function's contract above the function declaration. That way you don't need the explicit `Promise.resolve()` and you don't need to read to the end of the function to understand the function's contract.

Reply:
You're right -- I wasn't thinking about the consequences of being in an action when I made my async/await suggestions.